### ISSUE TITLE
exposed_derived_secrets dramatically increases hash_tree_root time

### ISSUE THREAD

**Issue by CarlBeek** [ethereum/eth2.0-specs#1110]
This issue is for visibility purposes. It is not a problem per se as it can be handled with intelligent caches, but its impact on test running-times is so large that I feel it is worth noting. (Also the caches in the tests are not this intelligent yet.)

In phase 1, an `exposed_derived_secrets` vector is added to the BeaconState:
`'exposed_derived_secrets': [['uint64'], EARLY_DERIVED_SECRET_PENALTY_MAX_FUTURE_EPOCHS]`, where `EARLY_DERIVED_SECRET_PENALTY_MAX_FUTURE_EPOCHS = 2**14`. 

Basically, this list is huge and pushes tests from ~55 second run time to ~300 seconds. (This can be partially mitigated by reducing `2**14` to `2**12` in the constant_presets, but `exposed_derived_secrets` still takes long to be list-comprehended.)

**Comment by dankrad** [comment]
Huh. Is this dominated by the hashing time? Would this be fixed by reintroducing the hash cache?

**Comment by CarlBeek** [comment]
Nope, this is with a hash cache implemented. The majority of the time is spent doing list comprehension in `hash_tree_root`.

**Comment by dankrad** [comment]
At the danger of introducing a kitchen sink into SSZ, I wonder if it is feasible to have a dict structure at reasonable cost/complexity... that would remove the need for a very long static list to store these values.

**Comment by CarlBeek** [comment]
>At the danger of introducing a kitchen sink into SSZ, I wonder if it is feasible to have a dict structure at reasonable cost/complexity... that would remove the need for a very long static list to store these values.

Hmm, interesting idea. While this would be the perfect use case and I think we could benefit in other places too, the added complexity may not be worth it.

Adding dicts also means we need to add the cleanup/pruning logic which is currently handled elegantly by circular arrays. 

This may be a time to revert to design goal 0:
> to minimize complexity, even at the cost of some losses in efficiency

Alternatively, we could also just rename SSZ to Complex SerialiZe.

**Comment by dankrad** [comment]
>    to minimize complexity, even at the cost of some losses in efficiency
I agree. I actually don't know what the most reasonable dict structure in a Merkle tree looks like, was just wondering if someone would contribute it. But yes, the current construction has a very elegent cleanup logic and losing it is probably not worth it.

**Comment by CarlBeek** [comment]
I am closing this issue as I believe it has served its purpose of bringing awareness to this problem.
